README.md
# eppz! (C# theme for Unity)


Distinct colors for `Unity` **classes** and **types**.

![alt](images/eppz-Code_1_727px_crop.png)

Carefully designed colors **with meanings**.

![alt](images/eppz-Code_2_727px.png)

Colored **code regions**: `using`, `namespace`, `attribute` and more.

![alt](images/eppz-Code_3_727px.png)

**JSON** coloring (distinct array / dictionary punctuation).

![alt](images/eppz-Code_6_727px.png)

**Markdown** coloring.

![alt](images/eppz-Code_5_727px.png)

**XML Documentation** coloring.

![alt](images/eppz-Code_4_727px.png)


### License

* [Creative Commons Attribution-NonCommercial 4.0 CC BY-NC](https://creativecommons.org/licenses/by-nc/4.0/legalcode)


### More

* Contributed by Geri Borbás [@_eppz](http://twitter.com/_eppz)
* Repository at [eppz/VSCode.Extension.eppz_Code](https://github.com/eppz/VSCode.Extension.eppz_Code.git)
* Theme is the C# adaptaion of the orginal iOS countepart [eppz/iOS.Library.eppz_xCode](https://github.com/eppz/iOS.Library.eppz_xCode)
* You can opt-out language usage analytics using *eppz-code.disableAnalytics* settings

CHANGELOG.md
# **eppz!** (C# theme for Unity)
## **Change log**


* **1.2.52** - *2020-07-23*

    + Extracted TypeScript support
    + Some formatting
    + HTML Example code

* **1.2.50** - *2020-07-23* (Html_support)

    + `.html` coloring (basic)
        + Tag name and tag attribute coloring
    + Contributed by EricOP (https://github.com/ericop)

* **1.2.41** - *2017-07-20*

    + Remove awesome animated `.svg` icon
        + It got unsupported in the meanwhile
    + Meta
        + Less review popup message

* **1.2.22** - *2017-05-01*

    + Awesome animated `.svg` icon

* **1.1.31** - *2017-04-22* (Json_support)

    + `.json` coloring
        + Array / Dictionary punctuations
        + Key-Value coloring
        + See more in [`eppz-code.json.json`](https://github.com/eppz/VSCode.Extension.eppz_Code/blob/master/themes/default/eppz-code.json.json)
    + `.ts` TypeScript support (basic)
        + Hotfixes only (full language not covered yet)
    + Meta
        + File extension analytics (what to support next)

* **1.0.42 - 1.1.25** - *2017-04-19* (Markdown_support)

    + `.md` markdown coloring
        + Headings, links, inline code, list identation levels and more
        + Punctuations toned down locally
        + See more in [`eppz-code.markdown.json`](https://github.com/eppz/VSCode.Extension.eppz_Code/blob/master/themes/default/eppz-code.markdown.json)
    + Fixed link / image coloring issues
    + Meta
        + Alternative review popup message
        + Basic usage analytics (language, review popup)
        + Marketplace media, metadata

* **1.0.36** - *2017-04-07*

    + Command `eppz.code.resetReviewCounters` (for resetting review notification counters)
    + "Review" renamed to "Rate"
    
* **1.0.35** - *2017-04-07*

    + Added a gentle reminder to rate the extension (that help others discover this theme)
        + 4 reminder at all in a lifetime (2nd, 5th, 200th, 2000th launch)

* **Features/Review_popup/0.0.3** - *2017-04-07*

    + Small tweaks on popup logic

* **Features/Review_popup/0.0.2** - *2017-04-07*

    + `Data`
        + Extension storage access
    + `ReviewPopup`
        + Shows review message (and determine whether it should be)

* **Features/Review_popup/0.0.1** - *2017-04-07*

    + Added Typescript extension bootstrap (no tests yet)
    + Fixed license SPDX expression

* **1.0.31** - *2017-04-06*

    + Resolved [issue #2](https://github.com/eppz/VSCode.Extension.eppz_Code/issues/2) that prevented C# extension to load IntelliSense properly
        + Contribution language `id` renamed to `csharp` (instead `cs`)

* **1.0.29** - *2017-04-05*

    + Marketplace media, metadata

* **1.0.21 - 1.0.26** - *2017-04-04*

    + Metadata, marketplace assets
    + Renamed (debug) extensions to use them as tags in marketplace

* **1.0.2** - *2017-04-04* (Release_candidate)

    + More sample code tweak
    + Hooked up Unity class definitions to `variable.other.object` in `member-access-expression` 
        + Unity classes colored right when accessing static class methods
    + Minor color tweaks

* **1.0.1** - *2017-04-04*

    + Tweaked sample codes for screenshots

* **1.0.0** - *2017-04-04*

    + More Unity classes
    + Hooked up scopes to theme

* **0.9.0** - *2017-04-04*

    + Added rules for Unity classes

* **0.8.0** - *2017-04-04*

    + XML Doc design / implementation

* **0.7.8** - *2017-04-04*

    + Theme settings swatches (base colors, highlights)

* **0.7.5** - 0.7.51 *2017-04-04*

    + Added new swatches to theme files

* **0.7.0** - *2017-04-04*

    + Removed theme variations
    + Refined colors

* **0.6.0** - *2017-04-03*

    + Implemented theme definitions
        + Implemented regions (with extended region scopes)
        + Finetuned design
    + Grammar tweaks
        + Don't extend `using-statement`

* **0.4.0** - *2017-04-02*

    + Implemented color design
        + Actually a color space focused towards important parts of the code

* **0.3.0** - *2017-04-02*

    + `type` extensions
        + Added some `UnityEngine` types to `support` scope domain (only for testing yet)

* **0.2.0** - *2017-04-01* (Grammar_stable)

    + Implemented every extension needed for theming
        + `attribute-section`    
        + `using-directive`
        + `using-statement`
        + `namespace-declaration`
    + New scope `regions` introduced
        + `regions.attribute`
        + `regions.using.directive`
        + `regions.using.statement`        
        + `regions.namespace`
        + `regions.namespace.declaration`
        + Inside regions still the official C# grammar is (re)applied

* **0.1.0** - *2017-04-01*

    + Extension reference chain (injections to the top of the rules)
        + Declare `C# (official)`
            + From official [Syntax grammar used for C# colorization](https://github.com/dotnet/csharp-tmLanguage)
            + Scope renamed to preserve rules (instead overwriting)
        + Declare `C# (extensions)`
            + Extension rule definitions
                + Fallback to `official` counterpart at the end of each rule
        + Declare `C# (extended)`
            + A patched version of `C# (official)`
                + Extended rules referened from `C# (extensions)`
        + Declare `C#`
            + An alias that handles `cs` files directly

* **0.0.4** - *2017-04-01*

    + Added some scope to theme
    + Factored out injections

* **0.0.2 - 0.0.3** - *2017-03-31*

    + Worked around including original C# grammar
        + Introduced a fake language only to be able to include rules from it
        + From official [Syntax grammar used for C# colorization](https://github.com/dotnet/csharp-tmLanguage)

* **0.0.1** - *2017-03-31*

    + Initial commit
    + Have something usable regarding reusing of official C# grammar rules
    + Combined pack of theme and grammar extensions


src/Data.ts
import * as vscode from 'vscode';
import { GoogleAnalytics } from './GoogleAnalytics';


export class Data
{


    // Singleton.
    private static instance: Data;
    static CreateInstanceWithContext(context: vscode.ExtensionContext)
    {
        Data.instance = new Data(context);
        Data.instance.launchCountSinceInstall++; // Increase launch counter
    }
    static Instance()
    { return Data.instance; }
    static Reset()
    {
        Data.instance.reviewDidClicked = false;
        Data.instance.launchCountSinceInstall = 0;
        Data.instance.analyticsHasDisabled = false;
        GoogleAnalytics.ReviewEvent('Reset');
    }


    context: vscode.ExtensionContext;


    private constructor(context: vscode.ExtensionContext)
    { this.context = context; }


    // `reviewDidClicked`.
    public get reviewDidClicked(): boolean
    { return this.context.globalState.get("reviewDidClicked") == true; }
    public set reviewDidClicked(value: boolean)
    { this.context.globalState.update("reviewDidClicked", value); }

    // `launchesSinceLastReviewClose`.
    public get launchCountSinceInstall(): number
    { return Number(this.context.globalState.get("launchCount")); }
    public set launchCountSinceInstall(value: number)
    { this.context.globalState.update("launchCount", value); }

    // `analyticsHasDisabled`.
    public get analyticsHasDisabled(): boolean
    { return this.context.globalState.get("analyticsHasDisabled") == true; }
    public set analyticsHasDisabled(value: boolean)
    { this.context.globalState.update("analyticsHasDisabled", value); }

    // `UUID`.
    public get UUID(): string
    {
        // Create lazy.
        if (this.context.globalState.get("UUID") == null)
        { this.context.globalState.update("UUID", this.GenerateUUID()); }
        return this.context.globalState.get("UUID") as string;
    }
    private GenerateUUID(): string
    {
        // From http://stackoverflow.com/questions/33005400/how-to-generate-a-uuid-in-visual-studio-code
        var hexValues = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C", "D", "E", "F"];

        // c.f. rfc4122 (UUID version 4 = xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx)
        var oct = "", tmp;
        for (var a = 0; a < 4; a++)
        {
            tmp = (4294967296 * Math.random()) | 0;
            oct += hexValues[tmp & 0xF] + hexValues[tmp >> 4 & 0xF] + hexValues[tmp >> 8 & 0xF] + hexValues[tmp >> 12 & 0xF] + hexValues[tmp >> 16 & 0xF] + hexValues[tmp >> 20 & 0xF] + hexValues[tmp >> 24 & 0xF] + hexValues[tmp >> 28 & 0xF];
        }

        // "Set the two most significant bits (bits 6 and 7) of the clock_seq_hi_and_reserved to zero and one, respectively"
        var clockSequenceHi = hexValues[8 + (Math.random() * 4) | 0];
        return oct.substr(0, 8) + "-" + oct.substr(9, 4) + "-4" + oct.substr(13, 3) + "-" + clockSequenceHi + oct.substr(16, 3) + "-" + oct.substr(19, 12);
    }
}

src/main.ts
'use strict';
import * as vscode from 'vscode';
import { Data } from './Data';
import { ReviewPopup } from './ReviewPopup';
import { GoogleAnalytics } from './GoogleAnalytics';


export function activate(context: vscode.ExtensionContext)
{
    // 👨‍💼 heavy business.
    Data.CreateInstanceWithContext(context);
    GoogleAnalytics.CreateInstanceWithContext(context);
    ReviewPopup.PopInContextIfNeeded(context);

    // 📊 Google Analytics.
    GoogleAnalytics.AppEvent("Launched");
    vscode.workspace.onDidOpenTextDocument((textDocument: vscode.TextDocument) =>
    {
        // Language / file extension statistics (investigating why people use `plaintext` language that much).
        var fileExtension = (textDocument.fileName.includes("."))
            ? textDocument.fileName.split('.').pop()
            : "not set";
        GoogleAnalytics.AppEvent(
            "Did Open Text Document",
            textDocument.languageId+" ("+fileExtension+")"
        );
    });    

    // 👉 direct invocations (for testing mainly).
    context.subscriptions.push(vscode.commands.registerCommand(
        'eppz.code.popUpReview', 
        () =>
        {
            Data.Instance().reviewDidClicked = false; // Reset
            ReviewPopup.PopInContext(context);
        }
        ));   
    context.subscriptions.push(vscode.commands.registerCommand(
        'eppz.code.resetReviewCounters',
        () => { Data.Reset(); }
        ));
}


export function deactivate() { }

src/Analytics.ts
import * as vscode from 'vscode';
import * as analytics from 'universal-analytics';


export class Analytics
{


    // Singleton.
    private static instance: Analytics;
    static CreateInstanceWithContext(context: vscode.ExtensionContext)
    { Analytics.instance = new Analytics(context); }
    static Instance()
    { return Analytics.instance; }

    // Settings.
    private static Disabled()
    { return (vscode.workspace.getConfiguration('eppz-code')['disableAnalytics'] == true); }


    context: vscode.ExtensionContext;
    visitor: any;
    

    private constructor(context: vscode.ExtensionContext)
    {
         this.context = context;
         if (Analytics.Disabled()) return;
         this.visitor = analytics('UA-37060479-24');
    }

    public static AppEvent(action: string, label: string = null)
    {
        if (Analytics.Disabled()) return;
        Analytics.instance.visitor.event('App', action, label).send();
    }

    public static ReviewEvent(action: string, label: string = null)
    {
        if (Analytics.Disabled()) return;
        Analytics.instance.visitor.event('Review', action, label).send();
    }
}

src/ReviewPopup.ts
import * as vscode from 'vscode';
import { Data } from './Data';
import { GoogleAnalytics } from './GoogleAnalytics';


export class ReviewPopup
{


    private static ShouldPop(): boolean
    {
        // Only if not clicked yet.
        if (Data.Instance().reviewDidClicked) return false;

        // Lookup gentle reminder launch counts.
        var should: boolean;
        [2, 35, 100, 2000].forEach((item) => should = should || (Data.Instance().launchCountSinceInstall == item));
        
        return should;    
    }

    public static PopInContextIfNeeded(context: vscode.ExtensionContext): void
    {
        if (ReviewPopup.ShouldPop() == false) return; // Only if should
        ReviewPopup.PopInContext(context);
    }

    public static PopInContext(context: vscode.ExtensionContext): void
    {
        // Literals.
        var messages =
        [
            {
                "text" : "Like **eppz!** (C# theme for Unity)? ✨⭐🌟⭐✨ Rate it on the Marketplace!",
                "label" : "Like?"
            },
            {
                "text" : "Help others discover **eppz!** (C# theme for Unity)! ✨⭐🌟⭐✨ Rate it on the Marketplace!",
                "label" : "Help others!"
            },
        ];
        var message = messages[Math.floor(Math.random() * messages.length)]; // Random question from above
        var url = "https://marketplace.visualstudio.com/items?itemName=eppz.eppz-code"; // Skip `#review-details` for now
        var uri = vscode.Uri.parse(url);
        
        // Show.       
        vscode.window.showInformationMessage(message.text, "Rate🌟").then
        (
            selectedOption =>
            {
                if (selectedOption == "Rate🌟")
                {
                    vscode.commands.executeCommand('vscode.open', uri); // Open
                    Data.Instance().reviewDidClicked = true; // Save
                    GoogleAnalytics.ReviewEvent('Review clicked', message.label);
                }
                else
                {
                    GoogleAnalytics.ReviewEvent('Close clicked', message.label);
                }
            }
        );
        GoogleAnalytics.ReviewEvent('Pop up', message.label);
    }
}

src/GoogleAnalytics.ts
import * as vscode from 'vscode';
import Analytics from 'electron-google-analytics';
import { Data } from './Data';


export class GoogleAnalytics
{


    // Singleton.
    private static instance: GoogleAnalytics;
    static CreateInstanceWithContext(context: vscode.ExtensionContext)
    { GoogleAnalytics.instance = new GoogleAnalytics(context); }
    static Instance()
    { return GoogleAnalytics.instance; }

    // Settings.
    private static Disabled()
    { return (vscode.workspace.getConfiguration('eppz-code')['disableAnalytics'] == true); }


    context: vscode.ExtensionContext;
    tracker: Analytics;


    // Custom Dimension `UUID` (Index: 1, Scope: User).
    static CustomDimension_1(): string
    { return Data.Instance().UUID; }

    // Custom Dimension `Color Theme` (Index: 2, Scope: Hit).
    static CustomDimension_2(): string
    { return vscode.workspace.getConfiguration('workbench')['colorTheme']; }    
    

    private constructor(context: vscode.ExtensionContext)
    {
         this.context = context;
         this.tracker = new Analytics('UA-37060479-24');
    }

    private static IsEnabled(): Boolean
    {
        // If not disabled, return true (reset event).
        if (GoogleAnalytics.Disabled() == false)
        {
            Data.Instance().analyticsHasDisabled = false; // Reset
            return true;
        }

        // If disabled, and event sent, return false (do nothing).
        if (Data.Instance().analyticsHasDisabled == true) return false;

        // If disabled, return false (do send event).
        GoogleAnalytics._Event("App", "Disable analytics");
        Data.Instance().analyticsHasDisabled = true;
        return false;
    }

    public static AppEvent(action: string, label: string = null)
    {
        if (GoogleAnalytics.IsEnabled())
        { GoogleAnalytics._Event("App", action, label); }
    }

    public static ReviewEvent(action: string, label: string = null)
    {
        if (GoogleAnalytics.IsEnabled())
        { GoogleAnalytics._Event("Review", action, label); }
    }

    private static _Event(category: string, action: string, label: string = null)
    {
        let params =
        {
            ec: category,
            ea: action,
            el: label,
            cd1: GoogleAnalytics.CustomDimension_1(),
            cd2: GoogleAnalytics.CustomDimension_2()
        };
        GoogleAnalytics.instance.tracker.send('event', params, Data.Instance().UUID)
            .then((response) => { return response; })
            .catch((error) => { return error; });
    }
}